android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java

```diff
@@ -187,408 +245,303 @@ public void getContactsAsync(final ReadableMap options, final Promise promise) {
-          }
-          if (city != null) {
-            details.putString("city", city);
-          }
-          if (country != null) {
-            details.putString("country", country);
-          }
-          if (region != null) {
-            details.putString("region", region);
-          }
-          if (neighborhood != null) {
-            details.putString("neighborhood", neighborhood);
-          }
-          if (postcode != null) {
-            details.putString("postcode", postcode);
-          }
-          if (poBox != null) {
-            details.putString("poBox", poBox);
-          }
+  private boolean isMissingPermissions() {
+    return Build.VERSION.SDK_INT >= 23 &&
+        ContextCompat.checkSelfPermission(
+            getReactApplicationContext(),
+            Manifest.permission.READ_CONTACTS) != PackageManager.PERMISSION_GRANTED;
+  }
 
-          String label;
+  private static class Contact {
+    private String contactId;
+    private String displayName;
+    private String givenName = "";
+    private String middleName = "";
+    private String familyName = "";
+    private String prefix = "";
+    private String suffix = "";
+    private String phoneticFirstName = "";
+    private String phoneticMiddleName = "";
+    private String phoneticLastName = "";
+    private String company = "";
+    private String jobTitle ="";
+    private String department ="";
+    private String nickname = "";
+    private boolean hasPhoto = false;
+    private String photoUri;
+    private String note;
+    private List<Item> emails = new ArrayList<>();
+    private List<Item> phones = new ArrayList<>();
+    private List<Item> dates = new ArrayList<>();
+    private List<PostalAddressItem> postalAddresses = new ArrayList<>();
+    private List<ImAddressItem> imAddresses = new ArrayList<>();
+    private List<UrlAddressItem> urlAddresses = new ArrayList<>();
+    private List<RelationshipItem> relationships = new ArrayList<>();
+
+    public Contact(String contactId) {
+      this.contactId = contactId;
+    }
 
-          switch (type) {
-            case CommonDataKinds.StructuredPostal.TYPE_HOME:
-              label = "home";
-              break;
-            case CommonDataKinds.StructuredPostal.TYPE_OTHER:
-              label = "other";
-              break;
-            case CommonDataKinds.StructuredPostal.TYPE_WORK:
-              label = "work";
-              break;
-            case CommonDataKinds.StructuredPostal.TYPE_CUSTOM:
-              label = "custom";
-              break;
-            default:
-              label = "unknown";
-          }
+    // convert to react native object
+    public WritableMap toMap(Set<String> fieldSet) throws ParseException {
+      WritableMap contact = Arguments.createMap();
+      contact.putString("id", contactId);
+      contact.putString("name", !TextUtils.isEmpty(displayName) ? displayName : givenName + " " + familyName);
+      contact.putString("firstName", givenName);
```

Comment:
I think if any value is null/empty we den't want to put it in the response at all. 

So, initializing fields of this class is not necessary. Perhaps a method like `put to map if not null` would do?

Reply:
> I think if any value is null/empty we den't want to put it in the response at all.

i do agree however prioritized parity with [existing functionality](https://github.com/expo/expo/blob/master/android/expoview/src/main/java/abi26_0_0/host/exp/exponent/modules/api/ContactsModule.java#L202), which helps it's simpler for now as well :)

Reply:
did change for firstName, middleName, and lastName

Reply:
🤔 well, we don't put it if it is null: https://github.com/expo/expo/blob/master/android/expoview/src/main/java/abi26_0_0/host/exp/exponent/modules/api/ContactsModule.java#L221
or am I missing your point? :)

Reply:
hazah! you caught me :)

Reply:
check.